### DISCUSSION TITLE
Custom spacy Lemmatizer

### DISCUSSION THREAD

**Discussion by thalishsajeed** [explosion/spaCy#7102] [Help: Coding & Implementations] [answered]
I'd like to find some documentation regarding how to tune the behavior of the spacy's lemmatizer 
For example - Currently words like media and data get changed to medium/datum.

I'm interested in tweaking this to reflect how I want the lemmatizer to behave.

I understand that Spacy v2.2 did not have a configurable API and I am interested in knowing how spacy3 can be used for solving this use case

P.S - I am not interested in switching to a lookup based lemmatizer , I'm interested in tweaking the POS rule based lemmatizer behavior.

Does anyone have any tips on how to accomplish this

**Comment by adrianeboyd** [comment, marked as answer]
There is not a lot of documentation about the rule-based lemmatizer algorithm at this point. It's probably easiest to just dive into the code, which is not too complicated:

https://github.com/explosion/spaCy/blob/4188beda871b1e40eb8d02b8a787b6878c89717e/spacy/pipeline/lemmatizer.py#L170-L235

There are three tables stored in `lemmatizer.lookups` (`rules`, `exc`, and `index`) that provide POS-based rules, exceptions, and word lists. There's a "base form" check for forms that shouldn't need to be processed further, and if it's not a base form, it looks up exceptions and applies the rules to generate a list of possible lemmas. The rules are suffix substitution rules that generate possible lemmas and the word list is used to prioritize known vs. unknown lemmas.

A number of languages have custom `rule_lemmatize` methods that have slightly different backoff/default behavior (the French lemmatizer backs off to the lookup table, the Dutch lemmatizer only works on lowercase forms, etc.), so check for `lemmatizer.py` under `spacy/lang/*` to see some more examples.

The `rule_lemmatize` and other `{mode}_lemmatize` methods all return a list of possible lemmas, but the lemmatizer component itself only uses the first one because a `Doc` can only store one lemma per token.

To modify the lemmatizer behavior, you can extend the lemmatizer and implement a new `mode` with its own `{mode}_lemmatize` method or if the current algorithm is sufficient, you can modify the lemmatizer tables to change rules / exceptions / word lists to get the results you want.
